### PR TITLE
Skip to update pull time and pull count for scanner robot account

### LINKED ISSUE #14638: Wrong image pull times and tag retention removing not correct images
**Expected behavior and actual behavior:**
- Problem: Older pushed images were actually not pulled by user nor system, but the last "pull time" field shows strange time and it won't match with "last pull" time in second place. Our tag retention removes wrong images as older images are getting newer pull time
- Waiting solution: Image pull time will be fixed? and tag retention is not removing wrong images 

**Steps to reproduce the problem:**
- Checking pull times manually
-- Projects -> choose project -> choose repository:
-- Check "Pull Time" from the last column: "4/13/21, 2:34 AM"
-- Check "PULL TIME" when holding cursor on "tags" column tag name: "1/7/21, 2:19 PM"
-- Pull times don't match
Screenshot 2021-04-13 at 15 36 53

- Tag retention Dry run matches with "Pull Time" in last column: "2021/04/12 23:34:27"
`
|                                 Digest                                  |       Tag       | Kind  | Labels |     PushedTime      |     PulledTime      |     CreatedTime     | Retention |
|-------------------------------------------------------------------------|-----------------|-------|--------|---------------------|---------------------|---------------------|-----------|
| sha256:XXX | master-39       | image |        | 2021/04/12 14:31:00 | 2021/04/12 22:45:38 | 2021/04/12 14:31:00 | RETAIN    |
| sha256:XXX | master-38       | image |        | 2021/04/07 06:02:22 | 2021/04/12 22:59:29 | 2021/04/07 06:02:22 | RETAIN    |
| sha256:XXX | development-200 | image |        | 2021/04/07 05:58:36 | 2021/04/12 22:59:33 | 2021/04/07 05:58:36 | RETAIN    |
| sha256:XXX | master-36       | image |        | 2021/01/22 09:28:44 | 2021/04/12 23:33:28 | 2021/01/22 09:28:44 | DEL       |
| sha256:XXX | master-35       | image |        | 2021/01/15 13:31:57 | 2021/04/12 23:33:54 | 2021/01/15 13:31:57 | DEL       |
| sha256:XXX | master-34       | image |        | 2021/01/07 11:53:00 | 2021/04/12 23:34:27 | 2021/01/07 11:53:00 | RETAIN    |
`

- Checking Harbor logs for the specific image
-- Logs -> resource -> Type selected image -> last pull time matches with holding cursor on "tags" column tag name": "1/7/21, 2:19 PM"
Screenshot 2021-04-13 at 15 39 27

**Versions:**
- harbor version: 2.2.0
- docker engine version: 20.10.5, build 55c4c88
- docker-compose version: 1.23.2, build 1110ad01

**Additional context:**

- **Harbor config files:** 
- **Log files:**

### PR DESCRIPTION
**Pull request by stonezdj**
Fixes #14638

Signed-off-by: stonezdj <daojunz@vmware.com>

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/controller/event/handler/internal/artifact.go [hunk 1]
```diff
@@ -240,3 +246,45 @@ func (a *Handler) onPush(ctx context.Context, event *event.ArtifactEvent) error
...
+		log.Debugf("failed to get scanner robot account prefix for project %d, error: %v", projectID, err)
+		return false
+	}
+	return strings.HasPrefix(operator, prefix)
+}
+
+type scanRobotPrefixFunc func(ctx context.Context, projManager project.Manager, scannerCtl scanner.Controller, projectID int64) (string, error)
+
+// scanRobotPrefix get the scanner robot account prefix by project id
+func scanRobotPrefix(ctx context.Context, projManager project.Manager, scannerCtl scanner.Controller, projectID int64) (string, error) {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:269]: we should make sure that the creation and validation are using the same func to get the prefix.

File: src/controller/event/handler/internal/artifact.go [hunk 2]
```diff
@@ -240,3 +246,45 @@ func (a *Handler) onPush(ctx context.Context, event *event.ArtifactEvent) error
...
+		log.Debugf("failed to get scanner registration for project %d, error: %v", proj.ProjectID, err)
+		return "", err
+	}
+	if scan == nil {
+		return "", fmt.Errorf("scanner registration is nil")
+	}
+	if len(scan.Name) == 0 {
+		return "", fmt.Errorf("scanner name is empty")
+	}
+	return fmt.Sprintf("%s%s+%s", robotPrefix, proj.Name, scan.Name), nil
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:289]: and we could add a uuid, which is created on launching, for the naming pattern?

File: src/controller/event/handler/internal/artifact.go (hunk 1 above)
Comment (stonezdj) [reply src/controller/event/handler/internal/artifact.go]: ok. add uuid in prefix

File: api/v2.0/swagger.yaml [hunk 3]
```diff
@@ -8734,6 +8734,9 @@ definitions:
       skip_audit_log_database:
         $ref: '#/definitions/BoolConfigItem'
         description: Whether skip the audit log in database
+      scanner_skip_update_pulltime:
+        $ref: '#/definitions/BoolConfigItem'
+        description: Whether skip update the pull time for scanner
```
Comment (wy65701436) [inline api/v2.0/swagger.yaml:8739]: Whether to skip the pull time for updating the scanner

File: api/v2.0/swagger.yaml [hunk 4]
```diff
@@ -9008,6 +9011,11 @@ definitions:
         description: The session timeout for harbor, in minutes.
         x-omitempty: true
         x-isnullable: true
+      scanner_skip_update_pulltime:
+        type: boolean
+        description: Whether skip to update pull time for scanner
```
Comment (wy65701436) [inline api/v2.0/swagger.yaml:9016]: ditto

File: src/controller/event/handler/internal/artifact.go [hunk 5]
```diff
@@ -240,3 +244,21 @@ func (a *Handler) onPush(ctx context.Context, event *event.ArtifactEvent) error
 
 	return nil
 }
+
+// isScannerUser check if the current user is a scanner user by its prefix
+// usually a scanner user should be named like `robot$<projectName>+<Scanner UUID (8byte)>-<Scanner Name>-<UUID>`
+// verify it by the prefix `robot$<projectName>+<Scanner UUID (8byte)>`
+func isScannerUser(ctx context.Context, operator string, projectID int64) bool {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:251]: we can transmit the repository of into ArtifactEvent this function to avoid one DB call?

**Review by wy65701436 — APPROVED** [review]
lgtm

File: api/v2.0/swagger.yaml (hunk 3 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml:8739]: Update

File: src/controller/event/handler/internal/artifact.go (hunk 5 above)
Comment (stonezdj) [reply src/controller/event/handler/internal/artifact.go]: Updated

**Review by chlins — APPROVED** [review]
lgtm

**Pull request merged**